### ISSUE TITLE
Rule #42: Keep your ~~hands~~ variables to yourself

### ISSUE THREAD

**Issue by I-am-Erk** [CleverRaven/Cataclysm-DDA#53625]
### Is your feature request related to a problem? Please describe.

Implementation of [character switching](https://github.com/CleverRaven/Cataclysm-DDA/issues/53523) is going to aggravate a previously known [issue with NPCs "knowing" you when you return as a new character](https://github.com/CleverRaven/Cataclysm-DDA/issues/26126). We now actually have most of the infrastructure to deal with this, and can solve it with what I hope is a slight modification of the dialogue variable system, followed by a less-slight-but-not-impossible review of use of variables throughout the dialogue code.

We currently track "npc_has_var" and "u_has_var". We use "u" as a form of global variable under the increasingly incorrect assumption that the user won't change, while "npc" is local and applies only in this conversation.

### Solution you would like.

We need a couple more versions of our existing variable functions:

`fac_has_var`: `fac` related variable functions would apply the variable to the *player faction*. Different new characters in the same world should have a different player faction, and so would not share these variables. However, NPCs within your faction would, so if you take over an NPC, they maintain this variable.

`global_has_var`: This replaces the common ill-advised use of `u` as a global variable. If you need multiple NPCs to react to a change, and that reaction should be persistent across playthroughs, use `global`.

`party_has_var`: This option would attach a `u` scale variable to only your current followers. This might be useful for something somewhat secret or minor that your character may not tell the whole faction about, but that anyone currently travelling with you would know. This is the least important of the three, but if we want to do it right we should still include it so that it can get integrated with the whole setup.

All these sets of functions should have the full set of variable operations including variable timers and mathematics. At least for now, we probably don't need a faction or global version of any other dialogue conditionals.

The new faction trust mechanic should be checked out, and we should make sure it applies to the player faction, not the individual character, and not to new characters in the same world. Overall, the major change involved in this whole thing is likely going to be having a distinct player faction that resets on death or on character switch. That said, I don't think that it's as important to make sure your newly created characters get a new faction as it is to make sure your followers know what they should, and don't know what they shouldn't.

### Describe alternatives you have considered.

We could probably do something with flags on the variables themselves to indicate if they should persist after character death or not, but I think that makes an already very verbose system into a nightmare with no real usability help.

### Additional context

At some point we may want to look into things like `party_has_skill` and things to assess if anyone in your faction has a particular skill or attribute, facilitating party-based play by letting your followers pipe up if they know something, but I think that can wait a lot longer. Likewise, `party_add_effect` to apply a specific effect to all your current followers and a few similar things would be neat.

**Comment by Ramza13** [comment]
So global vars totally exist at the moment and can do timers and math and anything u vars can.  Right now Portal Storms are the only thing using them I think but at some point migrating over older code is doable.

The other two could basically be copies of global vars with an extra map to store multiple versions worth.  Half tempted to use this as an excuse to shrink dialog var names from needing 3 parts at the same time.

**Comment by I-am-Erk** [comment]
> Half tempted to use this as an excuse to shrink dialog var names from needing 3 parts at the same time.

Please do. The current fields should be optional. It drives me mad.

**Comment by PatrikLundell** [comment]
Many "personal" missions ought to be possible to be given to multiple characters, with whoever finishes the mission first clearing the mission for all of them. It doesn't make sense that I'd refuse to give the mission to kill the jabberwock to anyone but a single person, as I'm probably mostly interested in killing that critter dead. There might be some cases where there's some "honor" involved (e.g. vs a betrothed, should that become a possible relation) that might allow someone to keep trying until giving up or dying and only then hand it over to someone else, but that's probably an exception.
If the same mission can be given to multiple people, there's probably a need to separate mission states into a quest giver part and a quester part to allow the questers to progress to different degrees in parallel.

**Comment by I-am-Erk** [comment]
That's mainly a separate issue. I'm trying to avoid delving too far into missions here because that code is a mess and a half

**Comment by I-am-Erk** [comment]
I am pretty sure we have all these by now, in various forms.
